Handling invalid git repository and hidden folder

xNephe opened the pull request:
This avoid getting an error 500 on an invalid git repository, as discussed in #176 and #231.

- the autoreloading app ignores hidden folders in `KLAUS_REPOS_ROOT`

- invalid repository are displayed in red in the repos list alongside an error message, instead of triggering an error 500.

The index page with invalid repos look like this:
![2019-06-11-185356_530x466_scrot](https://user-images.githubusercontent.com/15261469/59389116-89714700-8d3b-11e9-82b2-c1004007f962.png)


I am terrible at designing things, so advice to improve the display of invalid repositories are welcome!

jelmer commented:
FWIW the code looks reasonable to me; the tests all pass except for python 3.7 (fixed in master).

jonashaag commented on klaus/templates/repo_list.html:
```diff
@@ -36,4 +36,17 @@ <h2>
   {% endfor %}
 </ul>
 
+{% if invalid_repos is defined and invalid_repos|length>0 %}
```
Is it ever undefined?

xNephe replied:
Not in the current code. I will remove the check.

jonashaag commented on klaus/__init__.py:
```diff
@@ -17,8 +18,9 @@ class Klaus(flask.Flask):
-        repo_objs = [FancyRepo(path) for path in repo_paths]
+        repo_objs, invalid_repos = self.load_repos(repo_paths)
         self.repos = dict((repo.name, repo) for repo in repo_objs)
+        self.invalid_repos = invalid_repos
```
Shouldn't we also use `.name` for invalid repos? It's based on the path only, so you could factor it out from `Repo.name` to e.g. `utils.something` and use it here as well.

xNephe replied:
I am not sure what you mean here. Are you suggesting something like `self.invalid_repos = dict((repo.name, repo) for repo in invalid_repos)`? I don't see the use of a dict here...

jonashaag replied:
True, at the moment we don’t need it since we’re not mapping invalid repos to urls. If we do that at some point we can change it to a dict. 

OTOH we could even stuff both types of repo objects into the same dict and have a check whenever we list or show repos.

xNephe replied:
I would rather keep two separated collections. I can't think of an operation we would want to apply to both valid and invalid repositories without distinction (even for display, we always want to separate valid and invalid repositories...). Having them in the same collection would requires to split the collection whenever we iterate through it, which is error prone.

jonashaag commented on klaus/views.py:
```diff
@@ -37,7 +37,8 @@ def repo_list():
         sort_key = lambda repo: (-(repo.get_last_updated_at() or -1), repo.name)
     repos = sorted(current_app.repos.values(), key=sort_key)
-    return render_template('repo_list.html', repos=repos, base_href=None)
+    invalid_repos = sorted([os.path.basename(path) for path in current_app.invalid_repos])
```
See comment about `.name` above.

Maybe it's handy to use a `InvalidRepo` class that has a `.name` attribute. So `FancyRepo` for good repos, `InvalidRepo` for bad repos. Not sure.

xNephe replied:
An `InvalidRepo` class seems coherent to me, especially if more information about invalid repos is stored in the future (the reason why it is invalid for instance). I will add a new class for them.

jonashaag commented on tests/repos/scripts/hidden_folder:
```diff
@@ -0,0 +1,5 @@
+
+# Hide the directory
+cd ..
+mv hidden_folder .hidden_folder
```
Looks like some changes to the actual tests are missing from the commit?

xNephe replied:
No, this file add a hidden folder in `KLAUS_REPOS_ROOT`. The already existing test `test_wsgi_autoreload` will only pass if it correctly ignores hidden folders, but no new test case is needed for this feature.

xNephe replied:
However, I did not add a test for the invalid repo display part, I should create one.

jonashaag commented:
@xNephe you still working on this?

xNephe commented:
I didn't find time to work on it in the last weeks, I have been pretty busy, but I haven't forgot it.
I hope to find some time by the end of the week.

xNephe commented:
I eventually found some time to fix the PR:
- I added an InvalidRepo object to abstract invalid repos
- I added a test to ensure invalid repos are display in the repo list view

jonashaag commented:
Thanks a lot! Merged with a few minor changes in ee9d78044040b8bd9090b774838a9f6684941f2d and 7f3f5b49990076c0c0e5338a31a523a70c92b750

xNephe commented:
Thanks! However, I noticed you added a function `is_hidden_folder` that end up not being used, because of the switch from `listdir` to `glob.glob`. We may want to delete it?

jonashaag commented:
Oops thanks!

rotsix commented:
Hey, is there a way to hide those invalid repositories?
